verification/cbmc/proofs/default_cmm_generate_enc_materials/default_cmm_generate_enc_materials_harness.c

```diff
@@ -0,0 +1,116 @@
+             * generated data key, it must be the right length.
+             * The nondeterminism increases coverage.
+             */
+            if (nondet_bool()) unencrypted_data_key->len = props->data_key_len;
+            unencrypted_data_key->allocator = request_alloc;
+            __CPROVER_assume(aws_byte_buf_is_valid(unencrypted_data_key));
+        }
+    } else {
+        /*
+         * If the buffer is not NULL, the byte buffer must not have been modified;
+         * however, this allocation modifies the buffer and increases coverage.
+         */
+        if (nondet_bool()) ensure_byte_buf_has_allocated_buffer_member(unencrypted_data_key);
+    }
+    int ret;
+    return ret;
+}
+
+void default_cmm_generate_enc_materials_harness() {
+    /* Nondet input required to init cmm */
+    struct aws_cryptosdk_keyring *keyring        = malloc(sizeof(*keyring));
+    const struct aws_cryptosdk_keyring_vt vtable = { .vt_size    = sizeof(struct aws_cryptosdk_keyring_vt),
+                                                     .name       = ensure_c_str_is_allocated(SIZE_MAX),
+                                                     .destroy    = nondet_voidp(),
+                                                     .on_encrypt = nondet_bool() ? NULL : on_encrypt,
+                                                     .on_decrypt = nondet_voidp() };
+    /* Assumptions required to init cmm */
+    ensure_cryptosdk_keyring_has_allocated_members(keyring, &vtable);
+    __CPROVER_assume(aws_cryptosdk_keyring_is_valid(keyring));
+    __CPROVER_assume(keyring->vtable != NULL);
+
+    /* Nondet input */
+    struct aws_cryptosdk_cmm *cmm               = aws_cryptosdk_default_cmm_new(can_fail_allocator(), keyring);
+    struct aws_cryptosdk_enc_materials **output = malloc(sizeof(*output));
+    struct aws_cryptosdk_enc_request *request   = malloc(sizeof(*request));
+
+    /* Assumptions */
+    __CPROVER_assume(cmm != NULL);
+
+    __CPROVER_assume(output != NULL);
+
+    __CPROVER_assume(request != NULL);
+    request->alloc = can_fail_allocator();
+    __CPROVER_assume(aws_allocator_is_valid(request->alloc));
+    request->enc_ctx = malloc(sizeof(*request->enc_ctx));
+    __CPROVER_assume(request->enc_ctx != NULL);
+    ensure_allocated_hash_table(request->enc_ctx, MAX_NUM_ITEMS);
+    __CPROVER_assume(aws_cryptosdk_enc_request_is_valid(request));
+
+    enum aws_cryptosdk_alg_id alg_id;
+    if (nondet_bool()) request->requested_alg = alg_id;
+    // We can either add this OR add "if (!props) goto err;" after line 61 in the source code
+    // I think adding the check to the source code is better in this case.
+    // struct aws_cryptosdk_alg_properties *alg_props = aws_cryptosdk_alg_props(alg_id);
+    // __CPROVER_assume(aws_cryptosdk_alg_properties_is_valid(alg_props));
+
+    /* Nondet set self->default_alg to a valid alg_id */
+    __CPROVER_assume(aws_cryptosdk_default_cmm_set_alg_id(cmm, alg_id) == AWS_OP_SUCCESS);
```

Comment:
This misses the edge case where the CMM's `default_alg` is unset, i.e. it has the special value `DEFAULT_ALG_UNSET` (defined in `default_cmm.c`) which is not considered a valid algorithm: https://github.com/aws/aws-encryption-sdk-c/blob/d6ea33e9390960737d3c3888b014bbe229b42953/source/default_cmm.c#L176-L178

Reply:
Please see the comment below:

> Sorry, there were two things wrong in here:
> * `aws_cryptsodk_algorithm_is_known` was not defined (fixed by adding header.c to Makefile).
> * Using an the `cmm` initializer instead of a nondet. allocator (which I have included now).
> So this code has been removed.

Because of this, all cases are now covered.